### PR TITLE
Honour AS ratelimit settings for /login requests 

### LINKED ISSUE #8846: Application Services are rate-limited by IP address on /login
Since [MSC2778](https://github.com/matrix-org/matrix-doc/pull/2778)'s implementation application services are able to use `/login` to login as an interested user and create a device which aids them in performing end-to-end encryption.

However, `/login` has a ratelimit on it which is IP-based:

https://github.com/matrix-org/synapse/blob/514a240aedd972e6c8b707a7685c3d283f8618c9/synapse/rest/client/v1/login.py#L117-L126

This is obviously quite bad news for application services that need to login to a lot of user accounts on startup or somesuch. Currently there is no way to override this ratelimit.

As a solution, I propose only calling upon the ratelimiter for non `uk.half-shot.msc2778.login.application_service` identifier types (aka non-AS requests). If the request cannot be authenticated via an access token, it will immediately be rejected.

We may also want to move this line:

https://github.com/matrix-org/synapse/blob/514a240aedd972e6c8b707a7685c3d283f8618c9/synapse/rest/client/v1/login.py#L120

as someone could technically supply a huge JSON body over and over and upset the server that way.

### PR DESCRIPTION
**Pull request by erikjohnston**
I also fixed up some type hints here, as we do:

https://github.com/matrix-org/synapse/blob/d765e427f18ee59716db6b4e462510b8cce4a0e1/synapse/rest/client/v1/login.py#L255-L256

which would break if given non-async functions. These callbacks come from password provider code, e.g.:

https://github.com/matrix-org/synapse/blob/d765e427f18ee59716db6b4e462510b8cce4a0e1/synapse/handlers/auth.py#L1603-L1610

Fixes #8846.

### CODE REVIEW

File: synapse/rest/client/v1/login.py [hunk 1]
```diff
@@ -194,6 +204,7 @@ async def _complete_login(
         login_submission: JsonDict,
         callback: Optional[Callable[[Dict[str, str]], Awaitable[None]]] = None,
         create_non_existent_users: bool = False,
+        ratelimit=True,
```
Comment (clokep) [inline synapse/rest/client/v1/login.py:207]:
```suggestion
        ratelimit: bool = True,
```

**Comment by clokep** [comment]
Note that in #8890 I started consistently using the `maybe_awaitable` helper for things of this nature.

**Comment by erikjohnston** [comment]
> Note that in #8890 I started consistently using the `maybe_awaitable` helper for things of this nature.

*nods*, I didn't really want to make a code change here though

File: changelog.d/8920.bugfix [hunk 2]
```diff
@@ -0,0 +1 @@
+Fix login API to not be ratelimit application services that have ratelimiting disabled.
```
Comment (clokep) [inline changelog.d/8920.bugfix:1]:
```suggestion
Fix login API to not ratelimit application services that have ratelimiting disabled.
```

**Review by clokep — APPROVED** [review]
The code duplication of checking the rate limiter is unfortunate, but I can't think of a better way. LGTM.

**Comment by erikjohnston** [comment]
> The code duplication of checking the rate limiter is unfortunate, but I can't think of a better way. LGTM.

The vast majority of the time I spent on this PR was sullenly staring at my screen trying to figure out a better way of doing it.

**Pull request merged**